### PR TITLE
STONEBLD-1803 Add rpm-ostree task

### PR DESCRIPTION
**Pull request by stuartwdouglas**
This allows the multi platform controller to create images that are bootable on VMs.

### CODE REVIEW

**Comment by chmeliik** [comment]
Would you be able to explain more in the commit message? I don't really have any context for what this is and why we want it

File: task/rpm-ostree/0.1/rpm-ostree.yaml [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+    name: create-purl-sbom
+    resources: {}
+    script: |
+      #!/bin/python3
+      import json
+
+      with open("./sbom-cyclonedx.json") as f:
+        cyclonedx_sbom = json.load(f)
+
+      purls = [{"purl": component["purl"]} for component in cyclonedx_sbom["components"] if "purl" in component]
```
Comment (tkdchen) [inline task/rpm-ostree/0.1/rpm-ostree.yaml:159]: There was an update to this line for other build tasks: `cyclonedx_sbom.get("components", [])`

Comment (stuartwdouglas) [reply task/rpm-ostree/0.1/rpm-ostree.yaml]: Thanks, I have made the same change.

File: task/buildah-remote/0.1/buildah-remote.yaml [hunk 2]
```diff
@@ -104,7 +104,7 @@ spec:
   - env:
     - name: COMMIT_SHA
       value: $(params.COMMIT_SHA)
-    image: quay.io/redhat-user-workloads/rhtap-build-tenant/multi-arch-controller/hacktask-image-multi-platform-controller:build-dfd73-1698382902@sha256:d733531a07ede2e73500c500a33cdaf78332db146ffdc23d3c1a71b253a734b6
+    image: quay.io/redhat-appstudio/multi-platform-runner:01c7670e81d5120347cf0ad13372742489985e5f
```
Comment (chmeliik) [inline task/buildah-remote/0.1/buildah-remote.yaml:107]: IIUC the commit-hash tag for this image is pinned in the taskgen script, which means the image doesn't self-update (via Renovate). Updating the image would require bumping the tag in https://github.com/redhat-appstudio/multi-platform-controller/blob/main/cmd/taskgen/main.go, and bumping the taskgen image tag in `hack/generate-buildah-remote.sh`

Could be improved in the future?

File: task/rpm-ostree/0.1/rpm-ostree.yaml [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+    build.appstudio.redhat.com/build_type: rpm-ostree
+    build.appstudio.redhat.com/multi-platform-required: "true"
+  name: rpm-ostree
+spec:
+  description: RPM Ostree
+  params:
+  - description: Reference of the image rpm-ostree will produce.
+    name: IMAGE
+    type: string
+  - default: quay.io/sdouglas/rpm-ostree:dev
```
Comment (chmeliik) [inline task/rpm-ostree/0.1/rpm-ostree.yaml:18]: Assuming this is temporary

File: task/rpm-ostree/0.1/rpm-ostree.yaml [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+  - env:
+    - name: COMMIT_SHA
+      value: $(params.COMMIT_SHA)
+    image: quay.io/redhat-appstudio/multi-platform-runner:01c7670e81d5120347cf0ad13372742489985e5f
+    imagePullPolicy: Always
+    name: build
+    resources:
+      limits:
+        cpu: "2"
+        memory: 4402341478400m
```
Comment (chmeliik) [inline task/rpm-ostree/0.1/rpm-ostree.yaml:86]: If I'm reading it right, this is about 4 Gi? Could it be specified in a more readable way?

File: task/rpm-ostree/0.1/rpm-ostree.yaml [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+      chmod +x scripts/script-build.sh
+      rsync -ra scripts "$SSH_HOST:$BUILD_DIR"
+      ssh $SSH_ARGS "$SSH_HOST" podman run --mount type=bind,source=$BUILD_DIR/tmp,target=/var/tmp,relabel=shared --privileged -e CONTEXT="$CONTEXT" -e IMAGE_FILE="$IMAGE_FILE" -e IMAGE="$IMAGE" -e IMAGE_EXPIRES_AFTER="$IMAGE_EXPIRES_AFTER" -e COMMIT_SHA="$COMMIT_SHA"  --rm  -v "$BUILD_DIR/workspaces/source:$(workspaces.source.path):Z"  -v $BUILD_DIR/scripts:/script:Z --user=0  "$BUILDER_IMAGE" /script/script-build.sh
+      rsync -ra "$SSH_HOST:$BUILD_DIR/workspaces/source/" "$(workspaces.source.path)/"
+      cp -r rhtap-final-image /var/lib/containers/rhtap-final-image
+      buildah pull oci:rhtap-final-image
+      buildah images
+      buildah tag localhost/rhtap-final-image "$IMAGE"
+      container=$(buildah from --pull-never "$IMAGE")
+    securityContext:
```
Comment (chmeliik) [inline task/rpm-ostree/0.1/rpm-ostree.yaml:126]: Is the `container=...` command needed here?

File: task/rpm-ostree/0.1/rpm-ostree.yaml [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+  - description: Image repository where the built image was pushed
+    name: IMAGE_URL
+  - description: Digests of the base images used for build
+    name: BASE_IMAGES_DIGESTS
+  - description: The counting of Java components by publisher in JSON format
+    name: SBOM_JAVA_COMPONENTS_COUNT
+    type: string
+  - description: The Java dependencies that came from community sources such as Maven
+      central.
+    name: JAVA_COMMUNITY_DEPENDENCIES
```
Comment (chmeliik) [inline task/rpm-ostree/0.1/rpm-ostree.yaml:59]: Are these two needed? Is it that all the build tasks must declare the same results?

File: task/rpm-ostree/0.1/rpm-ostree.yaml [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+    workingDir: $(workspaces.source.path)
+  - image: quay.io/redhat-appstudio/syft:v0.94.0
+    name: sbom-syft-generate
+    resources:
+      limits:
+        memory: 6Gi
+      requests:
+        memory: 6Gi
+    script: |
+      syft oci-dir:/var/lib/containers/rhtap-final-image --file=$(workspaces.source.path)/sbom-cyclonedx.json --output=cyclonedx-json
```
Comment (chmeliik) [inline task/rpm-ostree/0.1/rpm-ostree.yaml:145]: `--file` is deprecated for this syft version https://github.com/redhat-appstudio/build-definitions/commit/327db8cc4fc4a28ae1aca08d441cf46b3a6beeb7

**Review by chmeliik — COMMENTED** [review]
Overall seems reasonable :+1: 

Got a couple details

File: hack/generate-buildah-remote.sh [hunk 8]
```diff
@@ -1,4 +1,4 @@
 #!/bin/bash
 
 SCRIPTDIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )"
-podman run -v "$SCRIPTDIR"/..:/data quay.io/redhat-user-workloads/rhtap-build-tenant/multi-arch-controller/multi-arch-controller:taskgen-221979523603b730feca730aeb9d43f24a3c1e67  --buildah-task=/data/task/buildah/0.1/buildah.yaml --remote-task=/data/task/buildah-remote/0.1/buildah-remote.yaml
+podman run -v "$SCRIPTDIR"/..:/data quay.io/redhat-user-workloads/rhtap-build-tenant/multi-arch-controller/multi-arch-controller:taskgen-e008b351d34597c08b6731f21de37b3103662431  --buildah-task=/data/task/buildah/0.1/buildah.yaml --remote-task=/data/task/buildah-remote/0.1/buildah-remote.yaml
```
Comment (chmeliik) [inline hack/generate-buildah-remote.sh:4]: This update (and the corresponding buildah-remote update) doesn't seem connected to the rest of this change. Could you split it into a separate commit?

File: task/rpm-ostree/0.1/rpm-ostree.yaml (hunk 3 above)
Comment (stuartwdouglas) [reply task/rpm-ostree/0.1/rpm-ostree.yaml]: Oops, I had been testing with a RHTAP built image in redhat-user-workloads, and missed the default was still personal. I will update it to the redhat-user-workloads one, but I can't just build on GH actions like other images as it needs to be multi platform image>

File: task/rpm-ostree/0.1/rpm-ostree.yaml (hunk 4 above)
Comment (stuartwdouglas) [reply task/rpm-ostree/0.1/rpm-ostree.yaml:86]: This was copied from the buildah-remote task originally, as I thought I could mostly keep them in sync. They are two different and this does not need 4g anyway, I will update.

File: task/rpm-ostree/0.1/rpm-ostree.yaml (hunk 5 above)
Comment (stuartwdouglas) [reply task/rpm-ostree/0.1/rpm-ostree.yaml:126]: no, it was a leftover from the copy

File: hack/generate-buildah-remote.sh (hunk 8 above)
Comment (stuartwdouglas) [reply hack/generate-buildah-remote.sh:4]: without this the buildah-remote task and this one would end up out of sync with the images they are using.

Comment (chmeliik) [reply hack/generate-buildah-remote.sh:4]: It doesn't seem like this script touches rpm-ostree though :thinking:

File: task/rpm-ostree/0.1/rpm-ostree.yaml (hunk 3 above)
Comment (chmeliik) [reply task/rpm-ostree/0.1/rpm-ostree.yaml]: Could be built in RHTAP eventually?

File: task/rpm-ostree/0.1/rpm-ostree.yaml (hunk 6 above)
Comment (stuartwdouglas) [reply task/rpm-ostree/0.1/rpm-ostree.yaml]: I have removed it. The original intention was to make it as close to the buildah task as possible, but it was not really possible.

File: task/rpm-ostree/0.1/rpm-ostree.yaml (hunk 7 above)
Comment (stuartwdouglas) [reply task/rpm-ostree/0.1/rpm-ostree.yaml]: fixed

File: hack/generate-buildah-remote.sh (hunk 8 above)
Comment (stuartwdouglas) [reply hack/generate-buildah-remote.sh]: Yea, there is another PR. It has been a long day.

Comment (chmeliik) [reply hack/generate-buildah-remote.sh]: Ah, ok. We can probably drop the change in the buildah-remote task as well then :slightly_smiling_face:

**Review by chmeliik — APPROVED** [review]
LGTM

Just the change in `task/buildah-remote/0.1/buildah-remote.yaml` is not necessary

**Comment by chmeliik** [comment]
> LGTM
> 
> Just the change in `task/buildah-remote/0.1/buildah-remote.yaml` is not necessary

Oh, and your CI check caught that anyway :tada:

File: task/buildah-remote/0.1/buildah-remote.yaml (hunk 2 above)
Comment (stuartwdouglas) [reply task/buildah-remote/0.1/buildah-remote.yaml]: It definitely can be improved.

**Comment by cgwalters** [comment]
> Would you be able to explain more in the commit message? I don't really have any context for what this is and why we want it

It's used by https://github.com/centos/sagano

See also https://github.com/containers/bootc and https://coreos.github.io/rpm-ostree/container/#creating-base-images

(At the current time the command requires privileges, but that's a bug we will eventually fix)
